File: src/java/playground/src/TwoSum.java
import java.util.*;
import java.util.stream.IntStream;

public class TwoSum {
    public static void main(String[] args) {
        List<Integer> values = Arrays.asList(255,7,11,55,15,17,6);
        var result = twoSumMap(
            values.stream().mapToInt(Integer::intValue).toArray(),
            13
        );

        result.forEach(i -> {
            System.out.println("index: " + i + ", value: " + values.get(i));
        });
    }

    private static List<Integer> twoSumMap(int[] nums, int target) {
        Map<Integer, Integer> numMap = new HashMap<>();
        
        for(int i = 0; i < nums.length; i++) {
            //if the target is equals the nums[i], return the list with the position (i)
            if (nums[i] == target) {
                return Arrays.asList(i);
            }

            //if not, calculate the complement
            //complement = how much do we need to reach the target from the current number?
            int complement = target - nums[i];

            if (numMap.containsKey(complement)) {
                return Arrays.asList(numMap.get(complement), i);
            }

            numMap.put(nums[i], i);
        }

        return new ArrayList<>();
    }
}

File: src/kotlin/playground/src/TwoSum.kt
fun main() {
    val values = listOf(255, 7, 11, 55, 15, 17, 6)

    val result = twoSumMap(nums = values.toIntArray(), target = 13)

    result.forEach { i ->
        println("index: $i, value: ${values[i]}")
    }
}

private fun twoSumMap(nums: IntArray, target: Int): List<Int> {
    val indexByValue = mutableMapOf<Int, Int>() // value -> index

    nums.forEachIndexed { i, value ->
        // If the target equals the current value, return only this
        // index (same behavior as your Java code)
        if (value == target) return listOf(i)

        val complement = target - value
        val j = indexByValue[complement]
        if (j != null) return listOf(j, i)

        indexByValue[value] = i
    }

    return emptyList()
}

/*
    [ Big-O ]
    Time: O(n) - percorre o array uma vez, sem loops aninhados
    Space: O(n) - usa um Map auxiliar que pode crescer até o tamanho do input
 */